### ISSUE TITLE
[spirv][vulkan] Reduce operation bitwidth after extension

### ISSUE THREAD

**Issue by kuhar** [iree-org/iree#12543]
When compiling code with inputs of small integer bitwidths, we often end up with early sign/zero-extensions followed by arithmetic operations. By either pushing the extensions to happel later or looking through vector insertions/extractions, we can maintain small integer bitwidths longer and emit ops over smaller integer types or go to fp32 without any loss of precision.

For example:
```mlir
%x = arith.extsi %arg0: vector<4xi8> to vector<4xi32>
%y = arith.extsi %arg1: vector<4xi8> to vector<4xi32>
...
// %a is derived from %x
// %b is derived from %y

%mul = arith.muli %a, %b: vector<4xi32>
%res = arith.addi %mul, %acc: vector<4xi32>

=>

...
// %a is derived from %arg0
// %b is derived from %arg1

%x = arith.extsi %a: vector<4xi8> to vector<4xi32>
%y = arith.extsi %b: vector<4xi8> to vector<4xi32>
%mul = arith.muli %x, %y: vector<4xi32>
%res = arith.addi %mul, %acc: vector<4xi32>

=> 

...
// %a is derived from %arg0
// %b is derived from %arg1

%x = arith.extsi %a: vector<4xi8> to vector<4xi16>
%y = arith.extsi %b: vector<4xi8> to vector<4xi16>
%mul = arith.muli %x, %y: vector<4xi16>
%ext = arith.extsi %mul: vector<4xi16> to vector<4xi32> 
%res = arith.addi %ext, %acc: vector<4xi32>

or
=>

...
// %a is derived from %arg0
// %b is derived from %arg1

%x = arith.sitofp %a: vector<4xi8> to vector<4xf32>
%y = arith.sitofp %b: vector<4xi8> to vector<4xf32>
%r = arith.sitofp %acc: vector<4xi32> to vector<4xf32>  // This needs preconditions on %acc.  
%fma = vector.fma %x, %y, %r : vector<4xf32>
%res = arith.fptosi %fma: vector<4xfp32> to vector<4xi32> 
```

This should help with i8 matmul codegen for mobile GPUs that do not support integer dot product and with quantized models in general.

I implemented this in uVkCompute and saw big uplifts in i8 matmul performance w/ both outerproduct and innerproduct lowering:
- https://github.com/google/uVkCompute/blob/main/benchmarks/matmul/matmul_tiled_i8.glsl
- https://github.com/google/uVkCompute/blob/main/benchmarks/matmul/matmul_tiled_i8_innerproduct.glsl

The majority of the implementation will go into MLIR. This is an umbrella bug for making the improvement happen in IREE and measuring it.

**Comment by benvanik** [comment]
Great writeup and really excited to see progress on this! We need it on all backends and if it can be done in a generic enough way we could apply it to entire programs. There's a bit of infra in the flow dialect today that may be applicable here but it's only implemented for floats: InferNumericNarrowing and OptimizeNumerics. If that could be extended to work with integer types it may sink/hide enough of the widening/narrowing that you see it much less in the backends (and we move less data across dispatch boundaries through global memory/etc).

**Comment by dcaballe** [comment]
This is really cool! As Ben mentioned, hopefully we can get some of this working for all the backends! I found similar cases in the past where this optimization could be useful. For example, i8 -> i32 extension followed by transpose op, i32 transpose followed by truncations or i8 -> i32 extension followed by store at the end of the dispatch. Also something to take into account is that some targets may not have (efficient) support for some operations on a specific data type. IIRC, some x86 CPUs didn't have vector support for i8 divisions and they were scalarized. AVX-512 also has quite a few sub-ISAs that are data type specific so it's worth taking that into account as well.

**Comment by benvanik** [comment]
Agreed - I see lots of those kind of patterns in inputs coming from basic quantization tools that start with quantize/dequantize ops and fail to propagate the resulting type mangling very far.

For the flow-level stuff we generally always want to go smaller no matter what the support is as we're optimizing for binary size, global transient memory consumption, and global memory bandwidth. Definitely depends on the use case and a favor-size or favor-speed flag would be useful. IME a good default is to do the narrowing then fall back to load 8bit > widen into 16bit to do the math > store back in 8bit since it dramatically improves all those metrics at the risk of an additional op or two in compute-bound workloads scaled 1:1 with the number of loads and stores... lol which would likely be similar looking to the ops we have today that this issue is about :P

**Comment by kuhar** [comment]
Another pattern from https://github.com/openxla/iree/issues/12708#issuecomment-1478737161:

```mlir
    %5 = linalg.index 1 : index
    %6 = arith.index_cast %5 : index to i64
    %7 = arith.sitofp %6 : i64 to f16
    %8 = arith.addf %7, %cst : f16
    %9 = arith.truncf %cst_0 : f32 to f16
```

**Comment by kuhar** [comment]
I talked with @matthias-springer about this and the first set of narrowing transformations that push down extension ops and make multiplication/addition/etc. is better implemented with simple rewrite patterns. These can go to the arith/vector dialects upstream.

For the last case with `linalg.index`, we could use the `ValueBounds` interface. It comes with helper functions that walk the IR backwards. Because there's no support for incremental computation / invalidation, it's not very useful in the case where modify the IR as push zero/sign extensions down.

**Comment by kuhar** [comment]
No planned anymore
